horovod/torch/mpi_ops.py

```diff
@@ -0,0 +1,341 @@
+# We keep input in order to make sure it does not get garbage collected
+# before the operation is finished.
+_handle_map = {}
+
+
+# Null pointer.
+_NULL = mpi_lib._ffi.NULL
+
+
+def _check_function(function_factory, tensor):
+    function = function_factory(tensor)
+    if not hasattr(mpi_lib, function):
+        raise ValueError('Tensor type %s is not supported.' % tensor.type())
+    if not tensor.is_contiguous():
+        raise ValueError('Tensor is required to be contiguous.')
+    return function
+
+
+def _allreduce_function_factory(tensor):
+    return 'horovod_torch_allreduce_async_' + tensor.type().replace('.', '_')
+
+
+def _allreduce_async(tensor, output, average, name):
+    function = _check_function(_allreduce_function_factory, tensor)
+    handle = getattr(mpi_lib, function)(tensor, output, average,
+                                        name.encode() if name is not None else _NULL)
+    _handle_map[handle] = (tensor, output)
+    return handle
+
+
+def allreduce_async(tensor, average=True, name=None):
+    """
+    A function that performs asynchronous averaging or summation of the input tensor
+    over all the Horovod processes. The input tensor is not modified.
+
+    The reduction operation is keyed by the name. If name is not provided, incremented
+    auto-generated name is used. The tensor type and shape must be the same on all
+    Horovod processes for a given name. The reduction will not start until all processes
+    are ready to send and receive the tensor.
+
+    Arguments:
+        tensor: A tensor to average and sum.
+        average: A flag indicating whether to compute average or summation,
+                 defaults to average.
+        name: A name of the reduction operation.
+
+    Returns:
+        A handle to the allreduce operation that can be used with `poll()` or
+        `synchronize()`.
+    """
+    output = tensor.new(tensor.shape)
+    return _allreduce_async(tensor, output, average, name)
+
+
+def allreduce(tensor, average=True, name=None):
+    """
+    A function that performs averaging or summation of the input tensor over all the
+    Horovod processes. The input tensor is not modified.
+
+    The reduction operation is keyed by the name. If name is not provided, incremented
+    auto-generated name is used. The tensor type and shape must be the same on all
+    Horovod processes for a given name. The reduction will not start until all processes
+    are ready to send and receive the tensor.
+
+    Arguments:
+        tensor: A tensor to average and sum.
+        average: A flag indicating whether to compute average or summation,
+                 defaults to average.
+        name: A name of the reduction operation.
+
+    Returns:
+        A tensor of the same shape and type as `tensor`, averaged or summed across all
+        processes.
+    """
+    handle = allreduce_async(tensor, average, name)
+    return synchronize(handle)
+
+
+def allreduce_async_(tensor, average=True, name=None):
```

Comment:
Why trailing underscore?

Reply:
Trailing underscore is common PyTorch naming convention for in-place operations.